kore/src/Kore/Log/DecidePredicateUnknown.hs

```diff
@@ -0,0 +1,135 @@
+    | ErrorInFilterMultiOr
+    | ErrorInTests
+    deriving stock (Show, Eq)
+
+data DecidePredicateUnknown = DecidePredicateUnknown
+    { action :: OnDecidePredicateUnknown
+    , smtLimit :: SMT.RetryLimit
+    , predicates :: NonEmpty (Predicate VariableName)
+    }
+    deriving stock (Show, Eq)
+
+instance Exception DecidePredicateUnknown where
+    toException = toException . SomeEntry
+    fromException exn =
+        fromException exn >>= fromEntry
+
+instance Debug DecidePredicateUnknown where
+    debugPrec w = \_ -> Pretty.pretty . show $ w
+
+instance Diff DecidePredicateUnknown where
+    diffPrec = diffPrecEq
+
+instance Pretty DecidePredicateUnknown where
+    pretty DecidePredicateUnknown{smtLimit = SMT.RetryLimit limit, predicates} =
+        Pretty.vsep
+            ( [ "Failed to decide predicate:"
+              , Pretty.indent 4 (pretty predicate)
+              ]
+                ++ do
+                    sideCondition <- sideConditions
+                    [ "with side condition:"
+                        , Pretty.indent 4 (pretty sideCondition)
+                        ]
+                ++ [ "SMT limit set at:"
+                   , Pretty.indent
+                        4
+                        ( case limit of
+                            Limit n -> pretty n
+                            Unlimited -> "infinity"
+                        )
+                   ]
+            )
+      where
+        predicate :| sideConditions = predicates
+
+instance Entry DecidePredicateUnknown where
+    entrySeverity DecidePredicateUnknown{action} =
+        case action of
+            WarnSimplificationEquationInApplication _ -> Warning
+            _ -> Error
+    contextDoc DecidePredicateUnknown{action} =
+        Just $
+            Pretty.align $
+                Pretty.vsep
+                    [ Pretty.hsep . catMaybes $
+                        [ Just "while applying equation"
+                        , (\loc -> Pretty.hsep ["at", pretty loc]) <$> case action of
+                            WarnSimplificationEquationInApplication loc -> loc
+                            ErrorInApplication loc -> loc
+                            _ -> Nothing
+                        ]
+                    , Pretty.hsep
+                        [ "in"
+                        , case action of
+                            WarnSimplificationEquationInApplication _ ->
+                                "Kore.Equation.Application.checkRequires"
+                            ErrorInApplication _ ->
+                                "Kore.Equation.Application.checkRequires"
+                            ErrorInMatchWith ->
+                                "Kore.Rewrite.Search.matchWith"
+                            ErrorInSimplifyClaimRule ->
+                                "Kore.Rewrite.Rule.Simplify.simplifyClaimRule"
+                            ErrorInFilterMultiOr ->
+                                "Kore.Rewrite.SMT.Evaluator.filterMultiOr"
```

Comment:
We'll have to keep these messages up to date somehow. I wonder if there's some way of making sure this code doesn't compile if the names change.

Reply:
I think this can be done with generics. Will investigate

Reply:
hmmm... we could make a dummy datatype in the same module as the function calls and then use generics to get the data type's module, but then we would have to remember to move the dummy type together with the functions, so it's not safe from re-factoring. A better way might be to use template haskell to get the module name....i think thats doable, but also, template haskell :/